### DISCUSSION TITLE
ib.placeOrder AssertionError

### DISCUSSION THREAD

**Discussion by marksandstrom** [erdewit/ib_insync#580] [Q&A]
Dear ib_insync community:

In recent days, in particular outside 9am-5pm ET, my ib_insync trading script tends to crash on ib.placeOrder AssertionError, per below:

`close_order LimitOrder(action='SELL', totalQuantity=1, lmtPrice=0.038, tif='GTC', eTradeOnly=False, firmQuoteOnly=False, account='__')

---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-79-81582f39cb68> in <module> [..]
    177             print(f"close_order {close_order}")
--> 178             close = ib.placeOrder(ac, close_order)
~\anaconda3\lib\site-packages\ib_insync\ib.py in placeOrder(self, contract, order)
    643         if trade:
    644             # this is a modification of an existing order
--> 645             assert trade.orderStatus.status not in OrderStatus.DoneStates
    646             logEntry = TradeLogEntry(now, trade.orderStatus.status, 'Modify')
    647             trade.log.append(logEntry)

AssertionError: 

`
The same code has worked fine previously, and still works reliably at least during liquid trading hours. There is nothing complex/advanced with the code placing the closing order. However, some of the attempted closing orders are made inactive by the broker due to margin requirements (which increase overnight), though mostly the order is just rejected, and the execution of the code resumes, without crashing, on the next line.

Regarding the ib_insync code comment # this is a modification of an existing order, the trading code only places an independent new order, and does not, itself, seek to modify any other/existing order.

How to get the code to return graciously from a (rejected?) placeOrder, and resume execution on the next line?

**Comment by erdewit** [comment]
> the trading code only places an independent new order.

Are you sure about that? Are you're not somehow reusing an older Order object?

**Reply by marksandstrom** [reply]
Before placing the (standalone) order there is a line that may adjust the limit:

`close_order.lmtPrice = lmt`

Might such pre-placement resetting of lmtPrice lead to the AssertionError, and why so?

Thanks for an explanation.

**Comment by erdewit** [comment]
Are you re-using the opening order object to send it as closing order?

**Reply by marksandstrom** [reply]
No, the close_order is formed standalone, per

`LimitOrder('SELL', 1, lmt, tif='GTC', account=account, eTradeOnly=False, firmQuoteOnly=False)`

**Comment by erdewit** [comment]
That new LimitOrder can't trigger the assert exception. The assert is raised when an already finished order is placed again.

**Comment by marksandstrom** [comment]
It may be that I didn't manually cancel via TWS each order for which the ib.placeOrder AssertionError occurred (before the code line for ib.cancelOrder would have gotten executed), such that there could have been an existing, non-cancelled GTC limit order placed for some cases where the AssertionError occurred.
